Essentials/src/main/java/com/earth2me/essentials/ProviderFactory.java

```diff
@@ -0,0 +1,128 @@
+import java.util.logging.Level;
+
+public class ProviderFactory {
+    private final Map<Class<? extends Provider>, Provider> providers = new HashMap<>();
+    private final Map<Class<? extends Provider>, List<Class<? extends Provider>>> registeredProviders = new HashMap<>();
+    private final Essentials essentials;
+
+    public ProviderFactory(final Essentials essentials) {
+        this.essentials = essentials;
+    }
+
+    /**
+     * Gets the provider which has been selected for the given type.
+     * @param providerClass The provider type.
+     * @return the provider or null if no provider could be selected for that type.
+     */
+    public <P extends Provider> P get(final Class<P> providerClass) {
+        final Provider provider = providers.get(providerClass);
+        if (provider == null) {
+            return null;
+        }
+        //noinspection unchecked
+        return (P) provider;
+    }
+
+    @SafeVarargs
+    public final void registerProvider(final Class<? extends Provider>... toRegister) {
+        for (final Class<? extends Provider> provider : toRegister) {
+            final Class<?> superclass = provider.getInterfaces().length > 0 ? provider.getInterfaces()[0] : provider.getSuperclass();
+            if (Provider.class.isAssignableFrom(superclass)) {
+                //noinspection unchecked
+                registeredProviders.computeIfAbsent((Class<? extends Provider>) superclass, k -> new ArrayList<>()).add(provider);
+                if (essentials.getSettings().isDebug()) {
+                    essentials.getLogger().info("Registered provider " + provider.getSimpleName() + " for " + superclass.getSimpleName());
+                }
+            }
+        }
+    }
+
+    public void finalizeRegistration() {
+        for (final Map.Entry<Class<? extends Provider>, List<Class<? extends Provider>>> entry : registeredProviders.entrySet()) {
+            final Class<? extends Provider> providerClass = entry.getKey();
+            Class<? extends Provider> highestProvider = null;
+            ProviderData highestProviderData = null;
+            int highestWeight = -1;
+            for (final Class<? extends Provider> provider : entry.getValue()) {
+                try {
+                    final ProviderData providerData = provider.getAnnotation(ProviderData.class);
+                    if (providerData.weight() > highestWeight && testProvider(provider)) {
+                        highestWeight = providerData.weight();
+                        highestProvider = provider;
+                        highestProviderData = providerData;
+                    }
+                } catch (final Exception e) {
+                    essentials.getLogger().log(Level.SEVERE, "Failed to initialize provider " + provider.getName(), e);
+                }
+            }
+            if (highestProvider == null) {
+                throw new IllegalStateException("No provider found for " + providerClass.getName());
```

Comment:
Doesn't this fail on Spigot for Paper-only providers?

Reply:
maybe